x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsSource.java

```diff
@@ -0,0 +1,147 @@
+        parser.declareString(ConstructingObjectParser.constructorArg(), INDEX);
+        parser.declareObject(ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> QueryProvider.fromXContent(p, ignoreUnknownFields, Messages.DATA_FRAME_ANALYTICS_BAD_QUERY_FORMAT), QUERY);
+        return parser;
+    }
+
+    private final String index;
+    private final QueryProvider queryProvider;
+
+    public DataFrameAnalyticsSource(String index, @Nullable QueryProvider queryProvider) {
+        this.index = ExceptionsHelper.requireNonNull(index, INDEX);
+        if (index.isEmpty()) {
+            throw ExceptionsHelper.badRequestException("[{}] must be non-empty", INDEX);
+        }
+        this.queryProvider = queryProvider == null ? QueryProvider.defaultQuery() : queryProvider;
+    }
+
+    public DataFrameAnalyticsSource(StreamInput in) throws IOException {
+        index = in.readString();
+        queryProvider = QueryProvider.fromStream(in);
+    }
+
+    public DataFrameAnalyticsSource(DataFrameAnalyticsSource other) {
+        this.index = other.index;
+        this.queryProvider = new QueryProvider(other.queryProvider);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(index);
+        queryProvider.writeTo(out);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDEX.getPreferredName(), index);
+        builder.field(QUERY.getPreferredName(), queryProvider.getQuery());
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == this) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        DataFrameAnalyticsSource other = (DataFrameAnalyticsSource) o;
+        return Objects.equals(index, other.index)
+            && Objects.equals(queryProvider, other.queryProvider);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(index, queryProvider);
+    }
+
+    public String getIndex() {
+        return index;
+    }
+
+    /**
+     * Get the fully parsed query from the semi-parsed stored {@code Map<String, Object>}
+     *
+     * @return Fully parsed query
+     */
+    public QueryBuilder getParsedQuery() {
+        Exception exception = queryProvider.getParsingException();
+        if (exception != null) {
+            if (exception instanceof RuntimeException) {
+                throw (RuntimeException) exception;
+            } else {
+                throw new ElasticsearchException(queryProvider.getParsingException());
+            }
+        }
+        return queryProvider.getParsedQuery();
+    }
+
+    Exception getQueryParsingException() {
+        return queryProvider == null ? null : queryProvider.getParsingException();
```

Comment:
it looks like `queryProvider` shouldn't be null as it is set to a default if null